bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/read/ScanAdapter.java

```diff
@@ -137,56 +138,97 @@ public void throwIfUnsupportedScan(Scan scan) {
-    return narrowRowSet(rowSet, scan.getFilter());
-  }
+  private void toByteStringRange(Scan scan, Query query) {
+    RangeSet<RowKeyWrapper> rangeSet = getRangeSet(scan);
 
-  private RowSet getRowSet(Scan scan) {
-    if (scan instanceof BigtableExtendedScan) {
-      return ((BigtableExtendedScan) scan).getRowSet();
-    } else {
-      RowSet.Builder rowSetBuilder = RowSet.newBuilder();
-      ByteString startRow = ByteString.copyFrom(scan.getStartRow());
-      if (scan.isGetScan()) {
-        rowSetBuilder.addRowKeys(startRow);
+    for (Range<RowKeyWrapper> guavaRange : rangeSet.asRanges()) {
+      if (guavaRange.hasLowerBound() && guavaRange.lowerBoundType() == BoundType.CLOSED
+          && guavaRange.hasUpperBound() && guavaRange.upperBoundType() == BoundType.CLOSED
+          && guavaRange.lowerEndpoint().equals(guavaRange.upperEndpoint())) {
+
+        query.rowKey(guavaRange.lowerEndpoint().getKey());
       } else {
-        RowRange.Builder range =  RowRange.newBuilder();
-        if (!startRow.isEmpty()) {
-          if (!OPEN_CLOSED_AVAILABLE || scan.includeStartRow()) {
-            // the default for start is closed
-            range.setStartKeyClosed(startRow);
-          } else {
-            range.setStartKeyOpen(startRow);
+        ByteStringRange byteRange = ByteStringRange.unbounded();
+
+        if (guavaRange.hasLowerBound()) {
+          switch (guavaRange.lowerBoundType()) {
+          case CLOSED:
+            byteRange.startClosed(guavaRange.lowerEndpoint().getKey());
+            break;
+          case OPEN:
+            byteRange.startOpen(guavaRange.lowerEndpoint().getKey());
+            break;
+          default:
+            throw new IllegalArgumentException(
+                "Unexpected lower bound type: " + guavaRange.lowerBoundType());
           }
         }
 
-        ByteString stopRow = ByteString.copyFrom(scan.getStopRow());
-        if (!stopRow.isEmpty()) {
-          if (!OPEN_CLOSED_AVAILABLE || !scan.includeStopRow()) {
-            // the default for stop is open
-            range.setEndKeyOpen(stopRow);
-          } else {
-            range.setEndKeyClosed(stopRow);
+        // handle end key
+        if (guavaRange.hasUpperBound()) {
+          switch (guavaRange.upperBoundType()) {
+          case CLOSED:
+            byteRange.endClosed(guavaRange.upperEndpoint().getKey());
+            break;
+          case OPEN:
+            byteRange.endOpen(guavaRange.upperEndpoint().getKey());
+            break;
+          default:
+            throw new IllegalArgumentException(
+                "Unexpected upper bound type: " + guavaRange.upperBoundType());
           }
         }
-        rowSetBuilder.addRowRanges(range);
+        query.range(byteRange);
+      }
+    }
+  }
+
+  private RangeSet<RowKeyWrapper> getRangeSet(Scan scan) {
+    if (scan instanceof BigtableExtendedScan) {
+      return rowRangeAdapter.rowSetToRangeSet(((BigtableExtendedScan) scan).getRowSet());
```

Comment:
Is this line valid?

Reply:
@ajaaym Does this seems valid?  I am planning to create PR for this much changes first.. as it already contains 18 Files changes.